### ISSUE TITLE
`reportUnnecessaryContains` does not understand equal values of different types

### ISSUE THREAD

**Issue by LEW21** [microsoft/pyright#6087]
**Describe the bug**
`reportUnnecessaryContains` reports that the `a in [b]` expression will evaluate to False, while it actually evaluates to True, when `a` and `b` have different types but are equal.

**Code or Screenshots**
```
$ cat a.py
from decimal import Decimal
print(5 in [Decimal('5.0')])

$ cat pyrightconfig.json 
{"reportUnnecessaryContains": true}

$ pyright a.py 
/home/linus/a.py
  /home/linus/a.py:3:7 - error: Expression will always evaluate to False since the types "Literal[5]" and "Decimal" have no overlap (reportUnnecessaryContains)
1 error, 0 warnings, 0 informations 

$ python a.py 
True
```

**VS Code extension or command-line**
Both as VS Code extension and as command-line tool. Versions 1.1.328 and 1.1.329.

**Comment by erictraut** [comment]
Yes, this is by design. There's no infallible way for a static type checker to determine whether the `in` operator will always produce a `True` or `False` result. This check therefore uses some heuristics. These heuristics tend to work the vast majority of the time, but they can produce false positives in rare cases like the one you cited. The problem is that any Python class can implement `__eq__` and `__ne__` in any way they want, and a static type checker will have no knowledge that they support values whose types have no overlap.

The `reportUnnecessaryContains` is not enabled by default in pyright. If you decide that it produces too many false positives in your code, you can disable it.

**Comment by LEW21** [comment]
On the other hand, `a == b` does not produce such errors in the strict mode. Shouldn't they be handled in the same way?

```python
#pyright: strict

from decimal import Decimal

print(5 == Decimal('5.0'))
print(5 in [Decimal('5.0')])
```

Expected: both produce errors, or none should produce errors.

**Comment by erictraut** [comment]
> Shouldn't they be handled in the same way?

No, the operators (`==` and `in`) are very different in how they work — and therefore need to be treated very differently by a type checker. The `reportUnnecessaryComparison` check applies only to cases where the LHS and RHS are both literal types. The `reportUnnecessaryContains` applies only in cases when the RHS is a builtin container class with known semantics, and it uses type analysis to determine whether the type of the LHS and the container element type have a type overlap (i.e. one is a subclass of another). If `reportUnnecessaryContains` were changed to apply only to literal types, it would be worthless. It wouldn't catch any of the bugs it is intended to target.

Both `reportUnnecessaryComparison` and `reportUnnecessaryContains` are off by default in "basic" type checking mode and enabled in strict mode. If you find that either of these doesn't meet your needs, you can choose to leave it disabled.
